central/image/datastore/store/postgres/store.go

```diff
@@ -857,35 +862,33 @@ func (s *storeImpl) Delete(ctx context.Context, id string) error {
 	}
 	defer release()
 
-	return s.deleteImageTree(ctx, conn, id)
-}
-
-func (s *storeImpl) deleteImageTree(ctx context.Context, conn *pgxpool.Conn, imageIDs ...string) error {
-	// Delete from image table.
-	if _, err := conn.Exec(ctx, "delete from "+imagesTable+" where Id = ANY($1::text[])", imageIDs); err != nil {
-		return err
-	}
-	// Get orphaned image components.
-	rows, err := s.db.Query(ctx, "select id from "+imageComponentsTable+" where not exists (select "+imageComponentsTable+".id FROM "+imageComponentsTable+", "+imageComponentEdgesTable+" WHERE "+imageComponentsTable+".id = "+imageComponentEdgesTable+".imagecomponentid)")
+	tx, err := conn.Begin(ctx)
 	if err != nil {
-		return pgutils.ErrNilIfNoRows(err)
+		return err
 	}
-	defer rows.Close()
-	var componentIDs []string
-	for rows.Next() {
-		var id string
-		if err := rows.Scan(&id); err != nil {
+
+	if err := s.deleteImageTree(ctx, tx, id); err != nil {
+		if err := tx.Rollback(ctx); err != nil {
 			return err
 		}
-		componentIDs = append(componentIDs, id)
+	}
+	return tx.Commit(ctx)
+}
+
+func (s *storeImpl) deleteImageTree(ctx context.Context, tx pgx.Tx, imageIDs ...string) error {
+	// Delete from image table.
+	if _, err := tx.Exec(ctx, "delete from "+imagesTable+" where Id = ANY($1::text[])", imageIDs); err != nil {
+		return err
 	}
 
 	// Delete orphaned image components.
-	if _, err := conn.Exec(ctx, "delete from "+imageComponentsTable+" where id = ANY($1::text[])", componentIDs); err != nil {
+	_, err := tx.Exec(ctx, "delete from "+imageComponentsTable+" where not exists (select "+imageComponentsTable+".id FROM "+imageComponentsTable+", "+imageComponentEdgesTable+" WHERE "+imageComponentsTable+".id = "+imageComponentEdgesTable+".imagecomponentid)")
+	if err != nil {
 		return err
 	}
+
 	// Delete orphaned cves.
-	if _, err := conn.Exec(ctx, "delete from "+imageCVEsTable+" where not exists (select "+imageCVEsTable+".id FROM "+imageCVEsTable+", "+componentCVEEdgesTable+" WHERE "+imageCVEsTable+".id = "+componentCVEEdgesTable+".imagecveid)"); err != nil {
+	if _, err := tx.Exec(ctx, "delete from "+imageCVEsTable+" where not exists (select "+imageCVEsTable+".id FROM "+imageCVEsTable+", "+componentCVEEdgesTable+" WHERE "+imageCVEsTable+".id = "+componentCVEEdgesTable+".imagecveid)"); err != nil {
```

Comment:
This one confuses me, shouldn't this just be 

`delete from image_cves where not exists (select imagecveid from imagecomponentcveedges)`